include/swift/Runtime/Metadata.h

```diff
@@ -2125,6 +2125,99 @@ struct GenericWitnessTable {
   void *PrivateData[swift::NumGenericMetadataPrivateDataWords];
 };
 
+/// The structure of a type metadata record.
+///
+/// This contains enough static information to recover type metadata from a
+/// name. It is only emitted for types that do not have an explicit protocol
+/// conformance record. 
+///
+/// This structure is notionally a subtype of a protocol conformance record
+/// but as we cannot change the conformance record layout we have to make do
+/// with some duplicated code.
+struct TypeMetadataRecord {
+private:
+  // Some description of the type that is resolvable at runtime.
+  union {
+    /// A direct reference to the metadata.
+    RelativeIndirectablePointer<Metadata> DirectType;
+    
+    /// An indirect reference to the metadata.
+    RelativeIndirectablePointer<const ClassMetadata *> IndirectClass;
```

Comment:
We support this by registering them from swift_getGenericMetadata(), although I suppose we could emit a reference to the type metadata accessor for the pattern instead, in case it hasn't been instantiated yet. Is that what you were suggesting? (If you mean patterns that are specialised dynamically by the name being looked up, I wasn't sure how to do this and also validate any protocol requirements on the type.)